packages/aws-cdk/lib/archive.ts

```diff
@@ -1,26 +1,39 @@
 import archiver = require('archiver');
 import crypto = require('crypto');
 import fs = require('fs-extra');
+import glob = require('glob');
+import path = require('path');
 
 export function zipDirectory(directory: string, outputFile: string): Promise<void> {
   return new Promise((ok, fail) => {
-    const output = fs.createWriteStream(outputFile);
-    const archive = archiver('zip');
     // The below options are needed to support following symlinks when building zip files:
-    // -  nodir: This will prevent symlinks themselves from being copied into the zip.
+    // - nodir: This will prevent symlinks themselves from being copied into the zip.
     // - follow: This will follow symlinks and copy the files within.
     const globOptions = {
       dot: true,
       nodir: true,
       follow: true,
-      cwd: directory
+      cwd: directory,
     };
-    archive.glob('**', globOptions);
-    archive.pipe(output);
-    archive.finalize();
+    const files = glob.sync('**', globOptions); // The output here is already sorted
+
+    const output = fs.createWriteStream(outputFile);
 
+    const archive = archiver('zip');
     archive.on('warning', fail);
     archive.on('error', fail);
+    archive.pipe(output);
+
+    files.forEach(file => { // Append files serially to ensure file order
+      const stream = fs.createReadStream(path.join(directory, file));
+      archive.append(stream, {
```

Comment:
Aren't NodeJS streams async? Are we 100% positive that after this statement, the stream has been fully read and added? Or are we sure that `.finalize()` will synchronously read and flush all streams? Do we need to await `finalize()`?

It concerns me that I don't understand the API fully, and the documentation doesn't seem to be very explicit about it.

Reply:
[From the example in the doc](https://github.com/archiverjs/node-archiver#quick-start): _finalize the archive (ie we are done appending files but streams have to finish yet)_

[From the doc](https://www.archiverjs.com/archiver#finalize): _Finalizes the instance and prevents further appending to the archive structure (queue will continue til drained)._

I've successfully created zips containing multiple files of more than 50MB with this code (resulting in zip file of several hundreds of MB). 

Looks like `finalize()` is a blocking statement.

Reply:
It's also possible to call `archive.file()` and let the library handle the read stream creation (need to check if the sort order is preserved across zips in this case).

Reply:
Tested `archive.file()` and order is not preserved across zips. If think that the only way to go here is to use `archive.append()` with `fs.createReadStream()`.

Reply:
Cool, yes you are right, I had missed this:

```ts
     output.once('close', () => ok());
```

I'm just very scared of accidental asyncness where you're not expecting it :(

Thanks, will merge when the validation build finishes.